profusion of redundant "overruled by outer forbid" messages

Consider this (erroneous) program:

```rust
#![forbid(unused)]

#[allow(unused)]
fn main() {
    println!("hello forbidden world");
}
```

As we expect, this fails to compile, because the outer "forbid" attribute contradicts the "allow" attribute. ("Forbid" is like "deny" except that it prevents itself from being subsequently overridden.) However, the error messages leave something to be desired:

```
$ rustc --version
rustc 1.18.0 (03fc9d622 2017-06-06)
$ rustc scratch.rs 
error[E0453]: allow(unused_imports) overruled by outer forbid(unused_imports)
 --> /home/ubuntu/rust/scratch.rs:3:9
  |
1 | #![forbid(unused)]
  |           ------ `forbid` level set here
2 | 
3 | #[allow(unused)]
  |         ^^^^^^ overruled by previous forbid

error[E0453]: allow(unused_variables) overruled by outer forbid(unused_variables)
 --> /home/ubuntu/rust/scratch.rs:3:9
  |
1 | #![forbid(unused)]
  |           ------ `forbid` level set here
2 | 
3 | #[allow(unused)]
  |         ^^^^^^ overruled by previous forbid

error[E0453]: allow(unused_assignments) overruled by outer forbid(unused_assignments)
 --> /home/ubuntu/rust/scratch.rs:3:9
  |
1 | #![forbid(unused)]
  |           ------ `forbid` level set here
2 | 
3 | #[allow(unused)]
  |         ^^^^^^ overruled by previous forbid

error[E0453]: allow(dead_code) overruled by outer forbid(dead_code)
 --> /home/ubuntu/rust/scratch.rs:3:9
  |
1 | #![forbid(unused)]
  |           ------ `forbid` level set here
2 | 
3 | #[allow(unused)]
  |         ^^^^^^ overruled by previous forbid

error[E0453]: allow(unused_mut) overruled by outer forbid(unused_mut)
 --> /home/ubuntu/rust/scratch.rs:3:9
  |
1 | #![forbid(unused)]
  |           ------ `forbid` level set here
2 | 
3 | #[allow(unused)]
  |         ^^^^^^ overruled by previous forbid

error[E0453]: allow(unreachable_code) overruled by outer forbid(unreachable_code)
 --> /home/ubuntu/rust/scratch.rs:3:9
  |
1 | #![forbid(unused)]
  |           ------ `forbid` level set here
2 | 
3 | #[allow(unused)]
  |         ^^^^^^ overruled by previous forbid

error[E0453]: allow(unreachable_patterns) overruled by outer forbid(unreachable_patterns)
 --> /home/ubuntu/rust/scratch.rs:3:9
  |
1 | #![forbid(unused)]
  |           ------ `forbid` level set here
2 | 
3 | #[allow(unused)]
  |         ^^^^^^ overruled by previous forbid

error[E0453]: allow(unused_must_use) overruled by outer forbid(unused_must_use)
 --> /home/ubuntu/rust/scratch.rs:3:9
  |
1 | #![forbid(unused)]
  |           ------ `forbid` level set here
2 | 
3 | #[allow(unused)]
  |         ^^^^^^ overruled by previous forbid

error[E0453]: allow(unused_unsafe) overruled by outer forbid(unused_unsafe)
 --> /home/ubuntu/rust/scratch.rs:3:9
  |
1 | #![forbid(unused)]
  |           ------ `forbid` level set here
2 | 
3 | #[allow(unused)]
  |         ^^^^^^ overruled by previous forbid

error[E0453]: allow(path_statements) overruled by outer forbid(path_statements)
 --> /home/ubuntu/rust/scratch.rs:3:9
  |
1 | #![forbid(unused)]
  |           ------ `forbid` level set here
2 | 
3 | #[allow(unused)]
  |         ^^^^^^ overruled by previous forbid

error[E0453]: allow(unused_attributes) overruled by outer forbid(unused_attributes)
 --> /home/ubuntu/rust/scratch.rs:3:9
  |
1 | #![forbid(unused)]
  |           ------ `forbid` level set here
2 | 
3 | #[allow(unused)]
  |         ^^^^^^ overruled by previous forbid

error: aborting due to 11 previous errors
```

We get a separate error for each lint in the "unused" lint group! This is kind of terrible! We would prefer to get a single error for the contradicting attributes, whose message names the offending attributes (it's weird to say "overruled by outer forbid(path_statements)", when the outer forbid is for _unused_, not _path_statements_).

A fix by the present author is forthcoming.

--- a/src/librustc/lint/context.rs
+++ b/src/librustc/lint/context.rs
@@ -291,16 +291,13 @@ impl LintStore {
         self.by_name.insert(name.into(), Removed(reason.into()));
     }
 
-    #[allow(unused_variables)]
-    fn find_lint(&self, lint_name: &str, sess: &Session, span: Option<Span>)
-                 -> Result<LintId, FindLintError>
-    {
+    fn find_lint(&self, lint_name: &str) -> Result<LintId, FindLintError> {
         match self.by_name.get(lint_name) {
             Some(&Id(lint_id)) => Ok(lint_id),
             Some(&Renamed(_, lint_id)) => {
                 Ok(lint_id)
             },
-            Some(&Removed(ref reason)) => {
+            Some(&Removed(_)) => {
                 Err(FindLintError::Removed)
             },
             None => Err(FindLintError::NotFound)
@@ -313,7 +310,7 @@ impl LintStore {
                                     &lint_name[..], level);
 
             let lint_flag_val = Symbol::intern(&lint_name);
-            match self.find_lint(&lint_name[..], sess, None) {
+            match self.find_lint(&lint_name[..]) {
                 Ok(lint_id) => self.levels.set(lint_id, (level, CommandLine(lint_flag_val))),
                 Err(FindLintError::Removed) => { }
                 Err(_) => {
@@ -731,7 +728,7 @@ pub trait LintContext<'tcx>: Sized {
                     continue;
                 }
                 Ok((lint_name, level, span)) => {
-                    match self.lints().find_lint(&lint_name.as_str(), &self.sess(), Some(span)) {
+                    match self.lints().find_lint(&lint_name.as_str()) {
                         Ok(lint_id) => vec![(lint_id, level, span)],
                         Err(FindLintError::NotFound) => {
                             match self.lints().lint_groups.get(&*lint_name.as_str()) {
@@ -1420,7 +1417,7 @@ impl Decodable for LintId {
     fn decode<D: Decoder>(d: &mut D) -> Result<LintId, D::Error> {
         let s = d.read_str()?;
         ty::tls::with(|tcx| {
-            match tcx.sess.lint_store.borrow().find_lint(&s, tcx.sess, None) {
+            match tcx.sess.lint_store.borrow().find_lint(&s) {
                 Ok(id) => Ok(id),
                 Err(_) => panic!("invalid lint-id `{}`", s),
             }

--- a/src/librustc/lint/context.rs
+++ b/src/librustc/lint/context.rs
@@ -721,7 +721,7 @@ pub trait LintContext<'tcx>: Sized {
         let mut pushed = 0;
 
         for result in gather_attrs(attrs) {
-            let v = match result {
+            let (is_group, lint_level_spans) = match result {
                 Err(span) => {
                     span_err!(self.sess(), span, E0452,
                               "malformed lint attribute");
@@ -729,13 +729,14 @@ pub trait LintContext<'tcx>: Sized {
                 }
                 Ok((lint_name, level, span)) => {
                     match self.lints().find_lint(&lint_name.as_str()) {
-                        Ok(lint_id) => vec![(lint_id, level, span)],
+                        Ok(lint_id) => (false, vec![(lint_id, level, span)]),
                         Err(FindLintError::NotFound) => {
                             match self.lints().lint_groups.get(&*lint_name.as_str()) {
-                                Some(&(ref v, _)) => v.iter()
+                                Some(&(ref v, _)) => (true,
+                                                      v.iter()
                                                       .map(|lint_id: &LintId|
                                                            (*lint_id, level, span))
-                                                      .collect(),
+                                                      .collect()),
                                 None => {
                                     // The lint or lint group doesn't exist.
                                     // This is an error, but it was handled
@@ -751,14 +752,18 @@ pub trait LintContext<'tcx>: Sized {
 
             let lint_attr_name = result.expect("lint attribute should be well-formed").0;
 
-            for (lint_id, level, span) in v {
+            for (lint_id, level, span) in lint_level_spans {
                 let (now, now_source) = self.lint_sess().get_source(lint_id);
                 if now == Forbid && level != Forbid {
-                    let lint_name = lint_id.to_string();
+                    let forbidden_lint_name = match now_source {
+                        LintSource::Default => lint_id.to_string(),
+                        LintSource::Node(name, _) => name.to_string(),
+                        LintSource::CommandLine(name) => name.to_string(),
+                    };
                     let mut diag_builder = struct_span_err!(self.sess(), span, E0453,
                                                             "{}({}) overruled by outer forbid({})",
-                                                            level.as_str(), lint_name,
-                                                            lint_name);
+                                                            level.as_str(), lint_attr_name,
+                                                            forbidden_lint_name);
                     diag_builder.span_label(span, "overruled by previous forbid");
                     match now_source {
                         LintSource::Default => &mut diag_builder,
@@ -769,7 +774,10 @@ pub trait LintContext<'tcx>: Sized {
                         LintSource::CommandLine(_) => {
                             diag_builder.note("`forbid` lint level was set on command line")
                         }
-                    }.emit()
+                    }.emit();
+                    if is_group { // don't set a separate error for every lint in the group
+                        break;
+                    }
                 } else if now != level {
                     let cx = self.lint_sess_mut();
                     cx.stack.push((lint_id, (now, now_source)));

--- a/src/test/ui/lint/outer-forbid.rs
+++ b/src/test/ui/lint/outer-forbid.rs
@@ -1,0 +1,22 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Forbidding a group (here, `unused`) overrules subsequent allowance of both
+// the group, and an individual lint in the group (here, `unused_variables`);
+// and, forbidding an individual lint (here, `non_snake_case`) overrules
+// subsequent allowance of a lint group containing it (here, `bad_style`). See
+// Issue #42873.
+
+#![forbid(unused, non_snake_case)]
+
+#[allow(unused, unused_variables, bad_style)]
+fn main() {
+    println!("hello forbidden world")
+}

--- a/src/test/ui/lint/outer-forbid.stderr
+++ b/src/test/ui/lint/outer-forbid.stderr
@@ -1,0 +1,29 @@
+error[E0453]: allow(unused) overruled by outer forbid(unused)
+  --> $DIR/outer-forbid.rs:19:9
+   |
+17 | #![forbid(unused, non_snake_case)]
+   |           ------ `forbid` level set here
+18 | 
+19 | #[allow(unused, unused_variables, bad_style)]
+   |         ^^^^^^ overruled by previous forbid
+
+error[E0453]: allow(unused_variables) overruled by outer forbid(unused)
+  --> $DIR/outer-forbid.rs:19:17
+   |
+17 | #![forbid(unused, non_snake_case)]
+   |           ------ `forbid` level set here
+18 | 
+19 | #[allow(unused, unused_variables, bad_style)]
+   |                 ^^^^^^^^^^^^^^^^ overruled by previous forbid
+
+error[E0453]: allow(bad_style) overruled by outer forbid(non_snake_case)
+  --> $DIR/outer-forbid.rs:19:35
+   |
+17 | #![forbid(unused, non_snake_case)]
+   |                   -------------- `forbid` level set here
+18 | 
+19 | #[allow(unused, unused_variables, bad_style)]
+   |                                   ^^^^^^^^^ overruled by previous forbid
+
+error: aborting due to previous error(s)
+

